src/lib/logical_query_plan/join_node.cpp

```diff
@@ -71,6 +71,57 @@ std::vector<std::shared_ptr<AbstractExpression>> JoinNode::output_expressions()
   return output_expressions;
 }
 
+std::shared_ptr<LQPUniqueConstraints> JoinNode::unique_constraints() const {
+  // Semi- and Anti-Joins act as mere filters for input_left().
+  // Therefore, existing constraints are forwarded as they remain valid.
+  if (join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue || join_mode == JoinMode::AntiNullAsFalse) {
+    return _forward_unique_constraints();
+  }
+
+  auto unique_constraints = std::make_shared<LQPUniqueConstraints>();
+
+  // No guarantees for Cross Joins and multi predicate joins
+  const auto predicates = join_predicates();
+  if (predicates.empty() || predicates.size() > 1) return unique_constraints;
+  // Also, no guarantees for Non-Equi/Theta-Joins
+  const auto join_predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicates().front());
+  if (!join_predicate || join_predicate->predicate_condition != PredicateCondition::Equals) return unique_constraints;
+
+  if (join_mode == JoinMode::Inner || join_mode == JoinMode::Left || join_mode == JoinMode::Right ||
+      join_mode == JoinMode::FullOuter) {
+    // Check uniqueness of join columns
+    bool left_operand_is_unique = left_input()->has_unique_constraint({join_predicate->left_operand()});
+    bool right_operand_is_unique = right_input()->has_unique_constraint({join_predicate->right_operand()});
+    if (left_operand_is_unique && right_operand_is_unique) {
+      // Due to the one-to-one relationship, the constraints of both sides remain valid.
+      const auto& left_unique_constraints = left_input()->unique_constraints();
+      const auto& right_unique_constraints = right_input()->unique_constraints();
+
+      for (auto it = left_unique_constraints->begin(); it != left_unique_constraints->end(); it++) {
+        unique_constraints->emplace_back(std::move(*it));
+      }
+      for (auto it = right_unique_constraints->begin(); it != right_unique_constraints->end(); it++) {
+        unique_constraints->emplace_back(std::move(*it));
+      }
+      return unique_constraints;
+
+    } else if (left_operand_is_unique) {
+      // Uniqueness on the left prevents duplication of records on the right
+      return right_input()->unique_constraints();
+
+    } else if (right_operand_is_unique) {
+      // Uniqueness on the right prevents duplication of records on the left
+      return left_input()->unique_constraints();
+
+    } else {
+      // No constraints to return.
+      return unique_constraints;
+    }
+  } else {
+    Fail("Unhandled JoinMode");
```

Comment:
Like so?

```cpp
    DebugAssert(false, "Unhandled JoinMode");
    return unique_constraints;
```

Reply:
No, instead of the if condition. `DebugAssert(!((join_mode == JoinMode::Inner || join_mode == JoinMode::Left || join_mode == JoinMode::Right || join_mode == JoinMode::FullOuter)), ...` (too lazy to resolve the negation, but you know what I mean). Same reason as for the previous one: Removes a level of indentation, keeping the reader's "stack" smaller.

Reply:
Ok, got it 👍